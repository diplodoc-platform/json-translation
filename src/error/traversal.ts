class TraversalError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'TraversalError';

        Error.captureStackTrace(this, this.constructor);
    }
}

export {TraversalError};
export default {TraversalError};
