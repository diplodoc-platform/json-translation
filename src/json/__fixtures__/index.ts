const withRefs = require('./with-refs.json');
const withoutRefs = require('./without-refs.json');
const pageConstructorData = require('./page-constructor-data.json');
const pageConstructorSchema = require('./page-constructor-schema.json');

export {withRefs, withoutRefs, pageConstructorData, pageConstructorSchema};
export default {withRefs, withoutRefs, pageConstructorData, pageConstructorSchema};
