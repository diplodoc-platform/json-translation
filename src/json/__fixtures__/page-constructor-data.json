{
  "menu": {
    "title": "DataLens Festival"
  },
  "blocks": [
    {
      "type": "header-block",
      "title": "Yandex DataLens Festival",
      "overtitle": "5–16 декабря",
      "description": "Онлайн-фестиваль Yandex Cloud для аналитиков и руководителей о лучших практиках визуализации и анализа данных с помощью Yandex DataLens. Эксперты помогут разобраться в нюансах работы с DataLens, сравнить сервис с другими решениями и запланировать переезд на новую BI‑систему. Участвуйте бесплатно.",
      "width": "s",
      "resetPaddings": true,
      "verticalOffset": "m",
      "theme": "default",
      "breadcrumbs": {
        "items": [
          {
            "url": "/",
            "text": "На главную"
          }
        ]
      },
      "background": {
        "image": {
          "mobile": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_cover_mob-new.png",
          "desktop": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_cover.png",
          "disableCompress": true
        },
        "color": "#EFF2F8",
        "fullWidth": false
      },
      "buttons": [
        {
          "text": "Хочу участвовать",
          "theme": "action",
          "size": "promo",
          "url": "#registration"
        }
      ]
    },
    {
      "type": "media-block",
      "direction": "media-content",
      "animated": false,
      "anchor": {
        "url": "1",
        "text": "Для кого"
      },
      "media": {
        "image": {
          "src": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_1.png",
          "disableCompress": true
        }
      },
      "disableShadow": true,
      "title": "Для кого этот фестиваль",
      "description": "* Для аналитиков, которые уже используют DataLens или хотят познакомиться с инструментом поближе\n* Для руководителей или тимлидов аналитики, которые выбирают BI‑систему для своей компании и хотят узнать о ней больше\"\n"
    },
    {
      "type": "media-block",
      "direction": "content-media",
      "animated": false,
      "title": "Вас ждут",
      "description": " * Видеоуроки и полезные материалы от BI‑экспертов Yandex Cloud, а также компаний «Ёлва», «ДелоТех» (Группа компаний «Дело») и других\n* Живые сессии вопросов и ответов, где вы сможете спросить экспертов обо всем\n* Практические задания от экспертов\n* Конкурс и сюрпризы для победителей",
      "media": {
        "image": {
          "src": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_2.png",
          "disableCompress": true
        }
      },
      "largeMedia": false,
      "disableShadow": true
    },
    {
      "type": "card-layout-block",
      "title": "Зачем участвовать",
      "anchor": {
        "url": "2",
        "text": "Зачем участвовать"
      },
      "colSizes": {
        "all": 12,
        "md": 3,
        "sm": 3
      },
      "children": [
        {
          "type": "card-with-image",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/step_01.svg",
          "title": "Новые навыки",
          "description": "Вы прокачаете свои хардскиллы в аналитике на практических кейсах"
        },
        {
          "type": "card-with-image",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/step_02.svg",
          "title": "Знания на базе опыта",
          "description": "Получите знания для принятия взвешенных решений о переезде на новую BI‑систему"
        },
        {
          "type": "card-with-image",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/step_03.svg",
          "title": "Комьюнити",
          "description": "Обменяетесь опытом в комьюнити единомышленников в Telegram"
        },
        {
          "type": "card-with-image",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/step_04.svg",
          "title": "Бонус",
          "description": "Сможете принять участие в конкурсе и получить приятный бонус"
        }
      ]
    },
    {
      "type": "content-layout-block",
      "textContent": null,
      "title": "Программа",
      "text": "Вы можете принять участие как в одном треке, так и в обоих. Онлайн‑фестиваль будет проходить на платформе Яндекс Практикума с доступом в личный кабинет. Участие бесплатное по предварительной регистрации."
    },
    {
      "type": "card-layout-block",
      "title": "hello",
      "anchor": {
        "url": "program",
        "text": "Программа",
        "title": "Неделя 1: создание дашбордов в Yandex DataLens",
        "children": [
          {
            "type": "card-with-image",
            "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_prog_1-new.png",
            "title": "Обзор Yandex DataLens и сравнение с другими системами",
            "description": " "
          },
          {
            "type": "card-with-image",
            "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_prog_2-new.png",
            "title": "Кейсы переноса отчётов на Yandex DataLens",
            "description": ""
          },
          {
            "type": "card-with-image",
            "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_prog_3.png",
            "title": "Сессия вопросов и ответов с экспертом",
            "description": ""
          }
        ]
      }
    },
    {
      "type": "card-layout-block",
      "title": "Неделя 2: корпоративная аналитика в Yandex Cloud",
      "children": [
        {
          "type": "card-with-image",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_prog_4-new.png",
          "title": "Что нужно знать об аналитике в облаке",
          "description": ""
        },
        {
          "type": "card-with-image",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_prog_5-new.png",
          "title": "Кейсы миграции аналитических систем на Yandex Cloud",
          "description": ""
        },
        {
          "type": "card-with-image",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_prog_6-new.png",
          "title": "Сессия вопросов и ответов c экспертом",
          "description": ""
        }
      ]
    },
    {
      "type": "slider-block",
      "title": "hello",
      "text": "Эксперты курса",
      "slidesToShow": {
        "md": 4,
        "lg": 4,
        "xl": 4
      },
      "children": [
        {
          "type": "card-with-image",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/speaker_photo_dubinin.png",
          "title": "Павел Дубинин",
          "description": "Менеджер по развитию бизнеса Yandex DataLens. Активный участник комьюнити Yandex DataLens в Telegram"
        },
        {
          "type": "card-with-image",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/speaker_photo_bunin.png",
          "title": "Роман Бунин",
          "description": "BI‑евангелист Yandex DataLens. Экс‑руководитель аналитики Yandex Go"
        },
        {
          "type": "card-with-image",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/speaker_photo_tereshkin.png",
          "title": "Дмитрий Терёшкин",
          "description": "Руководитель отдела анализа данных и отчётности компании «ДелоТех» (Группа компаний «Дело»)"
        },
        {
          "type": "card-with-image",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/speaker_photo_soloviev.png",
          "title": "Леонид Соловьев",
          "description": "Ведущий консультант компании «Ёлва»"
        }
      ]
    },
    {
      "type": "questions-block",
      "title": "Частые вопросы",
      "items": [
        {
          "title": "Кто может участвовать в фестивале? Что для этого нужно?",
          "text": "В фестивале могут участвовать все желающие аналитики, руководители и тимлиды из любой точки мира. Мероприятие проходит онлайн. Участие бесплатное."
        },
        {
          "title": "Нужна ли регистрация, чтобы слушать онлайн-уроки?",
          "text": "Мы рекомендуем зарегистрироваться, чтобы получить доступ ко всем материалам онлайн-фестиваля, а также не упустить возможность принять участие в конкурсе."
        },
        {
          "title": "Я не аналитик и не руководитель. Можно мне на фестиваль?",
          "text": "Программа ориентирована на аудиторию, которая знакома с принципами работы BI или имела опыт в построении системы бизнес-аналитики. В рамках этого мероприятия мы не будем рассматривать базовые понятия, но вы можете присоединиться к другим [мероприятиям Yandex Cloud](https://cloud.yandex.ru/events)."
        },
        {
          "title": "Что я получу после фестиваля?",
          "text": "Вы сможете  принять участие в конкурсе на лучшее выполнение задания. Сертификатов и дипломов за посещение мероприятий фестиваля не предусмотрено."
        },
        {
          "title": "Будут ли записи онлайн‑уроков?",
          "text": "Да, вы сможете посмотреть записи в личном кабинете и после окончания фестиваля — на [Youtube-канале Yandex Cloud](https://www.youtube.com/c/YandexCloudPlatform/)."
        }
      ]
    },
    {
      "type": "form-block",
      "anchor": {
        "url": "registration",
        "text": "Регистрация"
      },
      "textContent": null,
      "title": "Регистрируйтесь на фестиваль, <br> участие бесплатно",
      "slug": "datalens-fest-form"
    }
  ]
}
