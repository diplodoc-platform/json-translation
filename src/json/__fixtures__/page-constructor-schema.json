{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "naive page constructor schema",
    "properties": {
        "menu": {
            "type": "object",
            "properties": {
                "title": {
                    "type": "string",
                    "translatable": true
                }
            }
        },
        "blocks": {
            "type": "array",
            "items": {
                "type": "object",
                "anyOf": [
                    {"$ref": "#/definitions/block-with-title"},
                    {"$ref": "#/definitions/block-without-title"}
                ]
            }
        }
    },
    "definitions": {
        "block-with-title": {
            "type": "object",
            "additionalProperties": true,
            "properties": {
                "title": {
                    "type": "string",
                    "translatable": true
                },
                "items": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "title": {
                                "type": "string",
                                "translatable": true
                            },
                            "text": {
                                "type": "string",
                                "translatable": true
                            }
                        },
                        "additionalProperties": true
                    }
                }
            }
        },
        "block-without-title": {
            "type": "object",
            "additionalProperties": true
        }
    }
}
