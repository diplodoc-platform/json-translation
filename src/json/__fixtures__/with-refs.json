{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "properties": {
        "type": "object",
        "menu": {
            "$ref": "#/definitions/menu"
        }
    },
    "definitions": {
        "menu": {
            "type": "object",
            "additionalProperties": true
        }
    }
}
