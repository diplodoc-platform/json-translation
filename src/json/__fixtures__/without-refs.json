{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "properties": {
        "type": "object",
        "menu": {
            "type": "object",
            "additionalProperties": true
        }
    }
}
