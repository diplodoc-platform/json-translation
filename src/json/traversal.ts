import Ajv, {JSONType, CodeKeywordDefinition, KeywordCxt, _} from 'ajv';
import {JSONSchema7} from 'json-schema';

import {TraversalError} from 'src/error/traversal';

export interface JSONTraversalHandler<T> {
    handleNode(val: T): T;
}

type AjvCodeGenerationFunctionParameters<T> = {
    type: JSONType;
    handler: T;
};

export type JSONTraversalParameters = {
    schema: JSONSchema7;
    ajv?: Ajv;
};

class JSONTraversal {
    private ajv: Ajv;
    private schema: JSONSchema7;

    constructor(parameters: JSONTraversalParameters) {
        const {ajv, schema} = parameters;

        this.ajv = ajv ?? new Ajv();
        this.schema = schema;
    }

    traverse(json: object): void {
        let validate;
        let valid;

        try {
            validate = this.ajv.compile(this.schema);
            valid = validate(json);
        } catch (err: unknown) {
            let message = 'failed to traverse data';
            if (err instanceof Error) {
                message += ` ${err.message}`;
            }

            throw new TraversalError(message);
        }

        if (!valid) {
            throw new TraversalError(
                `data doesn't match schema ${JSON.stringify(validate.errors)}`,
            );
        }
    }

    addNodeHandler<T extends JSONType, U>(type: T, schemaKeyword: string, handler: U) {
        const keyword: CodeKeywordDefinition = {
            keyword: schemaKeyword,
            type,
            code: this.ajvCodeGenerationFunction({handler, type}),
        };

        this.ajv.addKeyword(keyword);
    }

    private ajvCodeGenerationFunction<T>(parameters: AjvCodeGenerationFunctionParameters<T>) {
        const {handler, type} = parameters;

        return function (cxt: KeywordCxt) {
            const {gen, data, it} = cxt;
            const {parentData, parentDataProperty} = it;

            gen.if(_`typeof ${data} == ${type} && ${parentData} !== undefined`, () => {
                const func = gen.scopeValue('func', {ref: handler});
                gen.assign(data, _`${func}(${data})`);
                gen.assign(_`${parentData}[${parentDataProperty}]`, data);
            });
        };
    }
}

export {JSONTraversal};
export default {JSONTraversal};
