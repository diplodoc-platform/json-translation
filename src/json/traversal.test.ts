import {JSONTraversal, JSONTraversalParameters} from './traversal';

import {pageConstructorData, pageConstructorSchema} from './__fixtures__';

describe('json traversal', () => {
    it('is able to modify data in place', () => {
        const data = JSON.parse(JSON.stringify(pageConstructorData));

        const traversalArguments: JSONTraversalParameters = {
            schema: pageConstructorSchema,
        };
        const traversal = new JSONTraversal(traversalArguments);
        traversal.addNodeHandler('string', 'translatable', () => 'i am able to modify strings');

        traversal.traverse(data);
        expect(data).toMatchSnapshot();
    });

    it('is able to collect values', () => {
        const data = JSON.parse(JSON.stringify(pageConstructorData));
        const strings: Array<string> = new Array<string>();

        const traversalArguments: JSONTraversalParameters = {
            schema: pageConstructorSchema,
        };
        const traversal = new JSONTraversal(traversalArguments);
        traversal.addNodeHandler('string', 'translatable', (s: string) => {
            strings.push(s);
            return s;
        });

        traversal.traverse(data);
        expect(strings).toMatchSnapshot();
    });
});
