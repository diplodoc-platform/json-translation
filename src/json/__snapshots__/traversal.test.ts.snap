// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json traversal is able to collect values 1`] = `
Array [
  "DataLens Festival",
  "Yandex DataLens Festival",
  "Для кого этот фестиваль",
  "Вас ждут",
  "Зачем участвовать",
  "Программа",
  "hello",
  "Неделя 2: корпоративная аналитика в Yandex Cloud",
  "hello",
  "Частые вопросы",
  "Кто может участвовать в фестивале? Что для этого нужно?",
  "В фестивале могут участвовать все желающие аналитики, руководители и тимлиды из любой точки мира. Мероприятие проходит онлайн. Участие бесплатное.",
  "Нужна ли регистрация, чтобы слушать онлайн-уроки?",
  "Мы рекомендуем зарегистрироваться, чтобы получить доступ ко всем материалам онлайн-фестиваля, а также не упустить возможность принять участие в конкурсе.",
  "Я не аналитик и не руководитель. Можно мне на фестиваль?",
  "Программа ориентирована на аудиторию, которая знакома с принципами работы BI или имела опыт в построении системы бизнес-аналитики. В рамках этого мероприятия мы не будем рассматривать базовые понятия, но вы можете присоединиться к другим [мероприятиям Yandex Cloud](https://cloud.yandex.ru/events).",
  "Что я получу после фестиваля?",
  "Вы сможете  принять участие в конкурсе на лучшее выполнение задания. Сертификатов и дипломов за посещение мероприятий фестиваля не предусмотрено.",
  "Будут ли записи онлайн‑уроков?",
  "Да, вы сможете посмотреть записи в личном кабинете и после окончания фестиваля — на [Youtube-канале Yandex Cloud](https://www.youtube.com/c/YandexCloudPlatform/).",
  "Регистрируйтесь на фестиваль, <br> участие бесплатно",
]
`;

exports[`json traversal is able to modify data in place 1`] = `
Object {
  "blocks": Array [
    Object {
      "background": Object {
        "color": "#EFF2F8",
        "fullWidth": false,
        "image": Object {
          "desktop": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_cover.png",
          "disableCompress": true,
          "mobile": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_cover_mob-new.png",
        },
      },
      "breadcrumbs": Object {
        "items": Array [
          Object {
            "text": "На главную",
            "url": "/",
          },
        ],
      },
      "buttons": Array [
        Object {
          "size": "promo",
          "text": "Хочу участвовать",
          "theme": "action",
          "url": "#registration",
        },
      ],
      "description": "Онлайн-фестиваль Yandex Cloud для аналитиков и руководителей о лучших практиках визуализации и анализа данных с помощью Yandex DataLens. Эксперты помогут разобраться в нюансах работы с DataLens, сравнить сервис с другими решениями и запланировать переезд на новую BI‑систему. Участвуйте бесплатно.",
      "overtitle": "5–16 декабря",
      "resetPaddings": true,
      "theme": "default",
      "title": "i am able to modify strings",
      "type": "header-block",
      "verticalOffset": "m",
      "width": "s",
    },
    Object {
      "anchor": Object {
        "text": "Для кого",
        "url": "1",
      },
      "animated": false,
      "description": "* Для аналитиков, которые уже используют DataLens или хотят познакомиться с инструментом поближе
* Для руководителей или тимлидов аналитики, которые выбирают BI‑систему для своей компании и хотят узнать о ней больше\\"
",
      "direction": "media-content",
      "disableShadow": true,
      "media": Object {
        "image": Object {
          "disableCompress": true,
          "src": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_1.png",
        },
      },
      "title": "i am able to modify strings",
      "type": "media-block",
    },
    Object {
      "animated": false,
      "description": " * Видеоуроки и полезные материалы от BI‑экспертов Yandex Cloud, а также компаний «Ёлва», «ДелоТех» (Группа компаний «Дело») и других
* Живые сессии вопросов и ответов, где вы сможете спросить экспертов обо всем
* Практические задания от экспертов
* Конкурс и сюрпризы для победителей",
      "direction": "content-media",
      "disableShadow": true,
      "largeMedia": false,
      "media": Object {
        "image": Object {
          "disableCompress": true,
          "src": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_2.png",
        },
      },
      "title": "i am able to modify strings",
      "type": "media-block",
    },
    Object {
      "anchor": Object {
        "text": "Зачем участвовать",
        "url": "2",
      },
      "children": Array [
        Object {
          "description": "Вы прокачаете свои хардскиллы в аналитике на практических кейсах",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/step_01.svg",
          "title": "Новые навыки",
          "type": "card-with-image",
        },
        Object {
          "description": "Получите знания для принятия взвешенных решений о переезде на новую BI‑систему",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/step_02.svg",
          "title": "Знания на базе опыта",
          "type": "card-with-image",
        },
        Object {
          "description": "Обменяетесь опытом в комьюнити единомышленников в Telegram",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/step_03.svg",
          "title": "Комьюнити",
          "type": "card-with-image",
        },
        Object {
          "description": "Сможете принять участие в конкурсе и получить приятный бонус",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/step_04.svg",
          "title": "Бонус",
          "type": "card-with-image",
        },
      ],
      "colSizes": Object {
        "all": 12,
        "md": 3,
        "sm": 3,
      },
      "title": "i am able to modify strings",
      "type": "card-layout-block",
    },
    Object {
      "text": "Вы можете принять участие как в одном треке, так и в обоих. Онлайн‑фестиваль будет проходить на платформе Яндекс Практикума с доступом в личный кабинет. Участие бесплатное по предварительной регистрации.",
      "textContent": null,
      "title": "i am able to modify strings",
      "type": "content-layout-block",
    },
    Object {
      "anchor": Object {
        "children": Array [
          Object {
            "description": " ",
            "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_prog_1-new.png",
            "title": "Обзор Yandex DataLens и сравнение с другими системами",
            "type": "card-with-image",
          },
          Object {
            "description": "",
            "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_prog_2-new.png",
            "title": "Кейсы переноса отчётов на Yandex DataLens",
            "type": "card-with-image",
          },
          Object {
            "description": "",
            "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_prog_3.png",
            "title": "Сессия вопросов и ответов с экспертом",
            "type": "card-with-image",
          },
        ],
        "text": "Программа",
        "title": "Неделя 1: создание дашбордов в Yandex DataLens",
        "url": "program",
      },
      "title": "i am able to modify strings",
      "type": "card-layout-block",
    },
    Object {
      "children": Array [
        Object {
          "description": "",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_prog_4-new.png",
          "title": "Что нужно знать об аналитике в облаке",
          "type": "card-with-image",
        },
        Object {
          "description": "",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_prog_5-new.png",
          "title": "Кейсы миграции аналитических систем на Yandex Cloud",
          "type": "card-with-image",
        },
        Object {
          "description": "",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/img_prog_6-new.png",
          "title": "Сессия вопросов и ответов c экспертом",
          "type": "card-with-image",
        },
      ],
      "title": "i am able to modify strings",
      "type": "card-layout-block",
    },
    Object {
      "children": Array [
        Object {
          "description": "Менеджер по развитию бизнеса Yandex DataLens. Активный участник комьюнити Yandex DataLens в Telegram",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/speaker_photo_dubinin.png",
          "title": "Павел Дубинин",
          "type": "card-with-image",
        },
        Object {
          "description": "BI‑евангелист Yandex DataLens. Экс‑руководитель аналитики Yandex Go",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/speaker_photo_bunin.png",
          "title": "Роман Бунин",
          "type": "card-with-image",
        },
        Object {
          "description": "Руководитель отдела анализа данных и отчётности компании «ДелоТех» (Группа компаний «Дело»)",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/speaker_photo_tereshkin.png",
          "title": "Дмитрий Терёшкин",
          "type": "card-with-image",
        },
        Object {
          "description": "Ведущий консультант компании «Ёлва»",
          "image": "https://storage.yandexcloud.net/cloud-www-assets/datalens-fest/speaker_photo_soloviev.png",
          "title": "Леонид Соловьев",
          "type": "card-with-image",
        },
      ],
      "slidesToShow": Object {
        "lg": 4,
        "md": 4,
        "xl": 4,
      },
      "text": "Эксперты курса",
      "title": "i am able to modify strings",
      "type": "slider-block",
    },
    Object {
      "items": Array [
        Object {
          "text": "i am able to modify strings",
          "title": "i am able to modify strings",
        },
        Object {
          "text": "i am able to modify strings",
          "title": "i am able to modify strings",
        },
        Object {
          "text": "i am able to modify strings",
          "title": "i am able to modify strings",
        },
        Object {
          "text": "i am able to modify strings",
          "title": "i am able to modify strings",
        },
        Object {
          "text": "i am able to modify strings",
          "title": "i am able to modify strings",
        },
      ],
      "title": "i am able to modify strings",
      "type": "questions-block",
    },
    Object {
      "anchor": Object {
        "text": "Регистрация",
        "url": "registration",
      },
      "slug": "datalens-fest-form",
      "textContent": null,
      "title": "i am able to modify strings",
      "type": "form-block",
    },
  ],
  "menu": Object {
    "title": "i am able to modify strings",
  },
}
`;
